File: frontend/src/api/dto/gpxPreviewDto.ts
import type { Instant } from './instant.ts'
import type { TrackDto } from './trackDto.ts'
import type { WaypointDto } from './waypointDto.ts'

/**
 * Analysed GPX file, not attached to any team
 */
export interface GpxPreviewDto {
  /** Public identifier used in URLs */
  id: string
  /** Track name */
  name: string
  /** Distance in meters */
  distance: number
  /** Total elevation gain in meters */
  elevationGain: number
  /** Total elevation loss in meters */
  elevationLoss: number
  /** Elevation gain per kilometer */
  hilliness: number
  /** Whether the current user created this preview */
  owned: boolean
  /** Creation timestamp */
  createdAt: Instant
  /** Rendered map thumbnail template URL (with a {size} placeholder), used for link previews; null when no thumbnail was generated */
  thumbnailUrl?: string
  /** URL to download the analysed GPX file */
  gpxUrl: string
  /** URL to download the analysed track as a FIT file */
  fitUrl: string
  /** Tracks */
  tracks: TrackDto[]
  /** Waypoints */
  waypoints: WaypointDto[]
}

File: backend/src/main/java/fr/pedalons/dto/gpx/response/GpxPreviewDto.java
package fr.pedalons.dto.gpx.response;

import static org.geolatte.geom.builder.DSL.g;
import static org.geolatte.geom.builder.DSL.point;
import static org.geolatte.geom.crs.CoordinateReferenceSystems.WGS84;

import fr.pedalons.domain.gpx.GpxPreview;
import fr.pedalons.dto.routes.response.TrackDto;
import fr.pedalons.dto.routes.response.WaypointDto;
import fr.pedalons.dto.validation.ValidateSchema;
import java.time.Instant;
import java.util.List;
import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.jspecify.annotations.Nullable;

/**
 * An analysed GPX file, shaped like the route detail payload so the frontend map and elevation
 * chart can render it unchanged.
 */
@Schema(description = "Analysed GPX file, not attached to any team")
@ValidateSchema
public record GpxPreviewDto(
    @Schema(description = "Public identifier used in URLs", required = true) String id,
    @Schema(description = "Track name", required = true) String name,
    @Schema(description = "Distance in meters", required = true) Float distance,
    @Schema(description = "Total elevation gain in meters", required = true) Float elevationGain,
    @Schema(description = "Total elevation loss in meters", required = true) Float elevationLoss,
    @Schema(description = "Elevation gain per kilometer", required = true) Float hilliness,
    @Schema(description = "Whether the current user created this preview", required = true)
        boolean owned,
    @Schema(description = "Creation timestamp", required = true) Instant createdAt,
    @Schema(
            description =
                "Rendered map thumbnail template URL (with a {size} placeholder), used for link"
                    + " previews; null when no thumbnail was generated")
        @Nullable String thumbnailUrl,
    @Schema(description = "Tracks", required = true) List<TrackDto> tracks,
    @Schema(description = "Waypoints", required = true) List<WaypointDto> waypoints) {

  public static GpxPreviewDto from(GpxPreview preview, boolean owned) {
    String publicId = preview.getPublicId().toString();
    return new GpxPreviewDto(
        publicId,
        preview.getName(),
        preview.getDistance(),
        preview.getElevationGain(),
        preview.getElevationLoss(),
        preview.getHilliness(),
        owned,
        preview.getCreatedAt(),
        preview.isHasThumbnail() ? "/api/gpx-previews/" + publicId + "/thumbnail/{size}" : null,
        preview.getTracks().stream().map(t -> TrackDto.of(t.trackPoints(), t.climbs())).toList(),
        preview.getWaypoints().stream()
            .map(w -> new WaypointDto(point(WGS84, g(w.lng(), w.lat())), w.name()))
            .toList());
  }
}
